ROX-25677: keep backward compatibility for duration format (#12329)
stackrox/stackrox: CHANGELOG.md

### Removed Features

### Deprecated Fatures

> - ROX-25677: The format for specifying durations in JSON requests will soon be restricted to a simpler format.

Reviewer: ```suggestion
- ROX-25677: The format for specifying durations in JSON requests will be restricted to a numeric format.
```

Author: Technically it's not numeric as it has `s` suffix

  Only a numeric value representing seconds (with optional fractional seconds for nanosecond precision)

>   followed by the s suffix will be accepted (e.g., "0.300s", "-5400s", or "9900").

Reviewer: ```suggestion
  followed by the s suffix will be accepted (e.g., "0.300s", "-5400s", or "9900s").
```

>   This replaces the current format, which allows a string with a potentially signed sequence of decimal numbers,

Reviewer: ```suggestion
  This replaces the current format, which allows a string with a signed sequence of decimal numbers,
```

  each with an optional fraction and a unit suffix (e.g., "300ms", "-1.5h", or "2h45m").

>   The currently valid time units "ns", "us" (or "µs"), "ms", "m", and "h" will no longer be supported in their existing form.

Reviewer 2: I will play devils-advocate:

Why would we go from a more feature-rich way of representing time, to a more limited way of representing time?

I will argue that we should take whatever technical measure we need to, to keep the option for specifying time as a string. We can't know how many customer scripts, workflows, cheatsheets, and automations may be relying on specifying time as a string. (We can change our own UI to seconds only, but our rule of thumb is that at least half our customer base uses our product without the UI, through roxctl or the API or some other integration.)

Author: Good point. This change affects only three endpoints related to CVE suppression (node, image, cluster). After switching to proto V2, the system became more strict about what it accepts, strictly adhering to proto conventions:

> [Generated output always contains 0, 3, 6, or 9 fractional digits, depending on required precision, followed by the suffix "s". Any fractional digits (or none) are accepted as long as they fit into nanoseconds precision, and the suffix "s" is required.](https://protobuf.dev/programming-guides/proto3/)

In this code, we use the legacy proto library – github.com/golang/protobuf/jsonpb, which is no longer maintained. Continuing to support it doesn't make sense beyond what's necessary.

Additionally, the current exception only applies to a single object. If we introduce a new API that uses duration in the future, we would need to extend our custom code to handle this new API object for consistency. It’s more practical to follow the established convention now.

>   The currently valid time units "ns", "us" (or "µs"), "ms", "m", and "h" will no longer be supported in their existing form.

Reviewer: ```suggestion
  The currently valid time units "ns", "us" (or "µs"), "ms", "m", and "h" will no longer be supported.
```

### Removed Features

### Deprecated Fatures
- ROX-25677: The format for specifying durations in JSON requests to
  `v1/nodecves/suppress`, `v1/clustercves/suppress` and `v1/imagecves/suppress`

>   will be restricted to a simpler format.

Author: ```suggestion
  will be restricted to a [proto JSON format](https://protobuf.dev/programming-guides/proto3/#json:~:text=are%20also%20accepted.-,Duration,-string).
```
